XCUITests/ScreenGraphTest.swift

```diff
@@ -0,0 +1,236 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import XCTest
+
+class ScreenGraphTest: XCTestCase {
+    var navigator: Navigator<TestUserState>!
+    var app: XCUIApplication!
+
+    override func setUp() {
+        app = XCUIApplication()
+        navigator = createTestGraph(for: self, with: app).navigator()
+        app.terminate()
+        restart(app, args: [LaunchArguments.ClearProfile, LaunchArguments.SkipIntro])
+    }
+
+    func restart(_ app: XCUIApplication, args: [String] = []) {
+        XCUIDevice.shared().press(.home)
+        var launchArguments = [LaunchArguments.Test]
+        args.forEach { arg in
+            launchArguments.append(arg)
+        }
+        app.launchArguments = launchArguments
+        app.activate()
+    }
+}
+
+extension ScreenGraphTest {
+    func testUserStateChanges() {
+        XCTAssertNil(navigator.userState.url, "Current url is empty")
+        navigator.goto(BrowserTab)
+
+        XCTAssertTrue(navigator.userState.url?.starts(with: "support.mozilla.org") ?? false, "Current url recorded by from the url bar")
+    }
+
+    func testBackStack() {
+        // We'll go through the browser tab, through the menu.
+        navigator.goto(SettingsScreen)
+        // Going back, there is no explicit way back to the browser tab,
+        // and the menu will have dismissed. We should be detecting the existence of
+        // elements as we go through each screen state, so if there are errors, they'll be
+        // reported in the graph below.
+        navigator.goto(BrowserTab)
+    }
+
+    func testSimpleToggleAction() {
+        // Switch night mode on, by toggling.
+        navigator.performAction(TestActions.ToggleNightMode)
+        XCTAssertTrue(navigator.userState.nightMode)
+        XCTAssertEqual(navigator.screenState, BrowserTab)
+
+        // Nothing should happen here, because night mode is already on.
+        navigator.toggleOn(navigator.userState.nightMode, withAction: TestActions.ToggleNightMode)
+        XCTAssertTrue(navigator.userState.nightMode)
+        XCTAssertEqual(navigator.screenState, BrowserTab)
+
+        // Switch night mode off.
+        navigator.toggleOff(navigator.userState.nightMode, withAction: TestActions.ToggleNightMode)
+        XCTAssertFalse(navigator.userState.nightMode)
+        XCTAssertEqual(navigator.screenState, BrowserTab)
+    }
+
+    func testChainedActionPerf1() {
+        let navigator = self.navigator!
+        measure {
```

Comment:
aside: I need to find out how we monitor this to indicate perf ok/bad

Reply:
:+1: